### PR TITLE
OAK-10025 - Documentation regarding conflicts is added.

### PR DESCRIPTION
**Pull request by jorgeflorez**
Documentation regarding conflicts is added.

### CODE REVIEW

File: oak-doc/src/site/markdown/dos_and_donts.md [hunk 1]
```diff
@@ -130,3 +130,19 @@ c = d.getParent();                              // preferred way to fetch the pa
...
 `InterruptibleChannel`, which are [asynchronously closed](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/channels/InterruptibleChannel.html) 
 when receiving an `InterruptedException` while blocked on IO. See [OAK-2609](https://issues.apache.org/jira/browse/OAK-2609).  
+
+### Avoid or minimize conflicts
+To reduce the possiblity of having errors like `OakState0001: Unresolved conflicts in ...`:
+
+1. Try to keep the JCR sessions as short as possible. i.e. create the session, make your changes, call session.save() and then session.logout(). If you need to do 
+something additional in the repository, a few lines after (maybe after some processing that could take some time), create the session again and repeat.
+
+2. Try to use session.refresh(true) before saving, if you think that some significant time can pass between the login() and the save() call.
```
Comment (anchela) [inline oak-doc/src/site/markdown/dos_and_donts.md]: i would recommend tocall session.refresh(false) before making changes to avoid conflicts. if that is not feasible, session.refresh(true) can be called before the save call.

File: oak-doc/src/site/markdown/dos_and_donts.md [hunk 2]
```diff
@@ -130,3 +130,19 @@ c = d.getParent();                              // preferred way to fetch the pa
 this is that Oak internally uses various classes from the `nio` package that implement 
 `InterruptibleChannel`, which are [asynchronously closed](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/channels/InterruptibleChannel.html) 
 when receiving an `InterruptedException` while blocked on IO. See [OAK-2609](https://issues.apache.org/jira/browse/OAK-2609).  
+
+### Avoid or minimize conflicts
+To reduce the possiblity of having errors like `OakState0001: Unresolved conflicts in ...`:
+
+1. Try to keep the JCR sessions as short as possible. i.e. create the session, make your changes, call session.save() and then session.logout(). If you need to do 
+something additional in the repository, a few lines after (maybe after some processing that could take some time), create the session again and repeat.
```
Comment (anchela) [inline oak-doc/src/site/markdown/dos_and_donts.md]: i am a bit reluctant to suggest to repeatedly create sessions. instead i would recommend calling Session.refresh(false) before starting with new changes after save() has been called.

File: oak-doc/src/site/markdown/dos_and_donts.md [hunk 3]
```diff
@@ -130,3 +130,19 @@ c = d.getParent();                              // preferred way to fetch the pa
 this is that Oak internally uses various classes from the `nio` package that implement 
 `InterruptibleChannel`, which are [asynchronously closed](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/channels/InterruptibleChannel.html) 
 when receiving an `InterruptedException` while blocked on IO. See [OAK-2609](https://issues.apache.org/jira/browse/OAK-2609).  
+
+### Avoid or minimize conflicts
+To reduce the possiblity of having errors like `OakState0001: Unresolved conflicts in ...`:
+
+1. Try to keep the JCR sessions as short as possible. i.e. create the session, make your changes, call session.save() and then session.logout(). If you need to do
```
Comment (anchela) [inline oak-doc/src/site/markdown/dos_and_donts.md]: this is correct. but i would rephrase it a bit. something like

make sure you always release the session by calling session.logout(). if possible avoid long-running sessions. if they are required (e.g. for observation) make sure to always call session.refresh(false) before applying changes or session.refresh(true) before saving the changes.

File: oak-doc/src/site/markdown/dos_and_donts.md [hunk 4]
```diff
@@ -130,3 +130,19 @@ c = d.getParent();                              // preferred way to fetch the pa
...
+something additional in the repository, a few lines after (maybe after some processing that could take some time), create the session again and repeat.
+
+2. Try to use session.refresh(true) before saving, if you think that some significant time can pass between the login() and the save() call.
+
+3. Enable the DEBUG level for `org.apache.jackrabbit.oak.plugins.commit.MergingNodeStateDiff` and `org.apache.jackrabbit.oak.plugins.commit.ConflictValidator` loggers if you want 
+to have more information on the circumstances of a conflict that happened in a point of time.
+
+4. Write your own conflict handler and add it when configuring your Oak or WhiteBoard instances. Only if you know what you are doing (i.e. you know how to resolve 
+the conflict in each one of the possible situations). By default, the [AnnotatingConflictHandler](https://jackrabbit.apache.org/oak/docs/apidocs/org/apache/jackrabbit/oak/plugins/commit/AnnotatingConflictHandler.html) instance will discard your changes 
+and your commit will fail. The worst that will happen is that your changes will not be saved (if you are ok with that).
```
Comment (anchela) [inline oak-doc/src/site/markdown/dos_and_donts.md]: here i would add:
- if persisting changes fails with a conflict and you cannot lose it, refactor your code such that you can retry after having called session.refresh(false)

File: oak-doc/src/site/markdown/dos_and_donts.md [hunk 5]
```diff
@@ -130,3 +130,19 @@ c = d.getParent();                              // preferred way to fetch the pa
...
+Please check the source code of [JcrLastModifiedConflictHandler](https://jackrabbit.apache.org/oak/docs/apidocs/org/apache/jackrabbit/oak/plugins/commit/JcrLastModifiedConflictHandler.html). It seems like a good example to follow.
```
Comment (anchela) [inline oak-doc/src/site/markdown/dos_and_donts.md]: 'it seems like a good example to follow' sounds a bit awkward in the oak documentation.
i would rephrase the point as follows:

Check the source code of [JcrLastModifiedConflictHandler](https://jackrabbit.apache.org/oak/docs/apidocs/org/apache/jackrabbit/oak/plugins/commit/JcrLastModifiedConflictHandler.html) for an example of a conflict handler.

**Review by anchela — CHANGES_REQUESTED** [review]
hi @jorgeflorez , thanks a lot for the PR. i very much like having conflicts covered by our best-practices section. there are just a few tiny adjustments i would make and left comments in the relevant sections.

File: oak-doc/src/site/markdown/dos_and_donts.md (hunk 5 above)
Comment (jorgeflorez) [reply oak-doc/src/site/markdown/dos_and_donts.md]: Hi @anchela, I think I got all changes you wanted. If I understood correctly this workflow, you can now view them :)

**Review by anchela — APPROVED** [review]
@jorgeflorez , lgtm

**Comment by anchela** [comment]
@jorgeflorez , thanks a lot for the contribution! i will merge it as soon as the check has completed.

**Comment by anchela** [comment]
merging without waiting for sonar checks (see OAK-10029)

File: oak-doc/src/site/markdown/dos_and_donts.md [hunk 6]
```diff
@@ -130,3 +130,17 @@ c = d.getParent();                              // preferred way to fetch the pa
...
 `InterruptibleChannel`, which are [asynchronously closed](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/channels/InterruptibleChannel.html) 
 when receiving an `InterruptedException` while blocked on IO. See [OAK-2609](https://issues.apache.org/jira/browse/OAK-2609).  
+
+### Avoid or minimize conflicts
+To reduce the possiblity of having errors like `OakState0001: Unresolved conflicts in ...`:
+
+1. Make sure you always release the session by calling session.logout(). If possible, avoid long-running sessions. If they are required (e.g. for observation) make sure 
+to always call session.refresh(false) before applying changes or session.refresh(true) before saving the changes. 
+
+2. Enable the DEBUG level for `org.apache.jackrabbit.oak.plugins.commit.MergingNodeStateDiff` and `org.apache.jackrabbit.oak.plugins.commit.ConflictValidator` loggers if you want
```
Comment (kwin) [inline oak-doc/src/site/markdown/dos_and_donts.md:140]: I don't think we should duplicate the info from https://jackrabbit.apache.org/oak/docs/FAQ.html#how-do-i-debug-merge-conflict-errors but rather add a link
